apollo-spaceport/src/lib.rs

```diff
@@ -0,0 +1,561 @@
+            match record {
+                StatsOrTrace::Stats(mut s) => {
+                    entry.stats_with_context.push(s.stats.take().unwrap())
+                }
+                StatsOrTrace::Trace(mut t) => entry.trace.push(t.trace.take().unwrap()),
+            }
+            // Drop the tpq lock to maximise concurrency
+            drop(tpq);
+
+            // This is inherently both imprecise and racy, but it doesn't matter
+            // because we are just hinting to the spaceport that it's probably a
+            // good idea to try to transfer data up to the ingress. Multiple
+            // notifications just trigger more transfers.
+            //
+            // 5000 is a fairly arbitrary number which indicates that we are adding
+            // a lot of data for transfer. It is intended to represent a rate of
+            // approx. 1,000 records/second.
+            let total = self.total.fetch_add(1, Ordering::SeqCst);
+            if total > 5000 {
+                let mut backoff = Duration::from_millis(0);
+                for i in 0..5 {
+                    match self.tx.send(()).await {
+                        Ok(_v) => {
+                            self.total.store(0, Ordering::SeqCst);
+                            break;
+                        }
+                        Err(e) => {
+                            tracing::warn!("attempt: {}, could not trigger transfer: {}", i + 1, e);
+                            if i == 4 {
+                                // Not being able to trigger a transfer isn't an "error". We can
+                                // let the client know that the transfer was Ok and hope that the
+                                // backend server eventually catches up with the workload and
+                                // clears the incoming trigger messages.
+                                break;
+                            } else {
+                                backoff += Duration::from_millis(50);
+                                tokio::time::sleep(backoff).await;
+                            }
+                        }
+                    }
+                }
+            }
+
+            let response = ReporterResponse {
+                message: "Report accepted".to_string(),
+            };
+            Ok(Response::new(response))
+        }
+    }
+
+    async fn extract_tpq(
+        client: &Client,
+        task_tpq: Arc<Mutex<HashMap<ReporterGraph, HashMap<String, report::TracesAndStats>>>>,
+    ) -> Vec<Result<Response<ReporterResponse>, Status>> {
+        let mut all_entries = task_tpq.lock().await;
+        let drained = all_entries
+            .drain()
+            .collect::<Vec<(ReporterGraph, HashMap<String, report::TracesAndStats>)>>();
+        // Release the lock ASAP so that clients can continue to add data
+        drop(all_entries);
+        let mut results = vec![];
+        for (graph, tpq) in drained {
+            if !tpq.is_empty() {
```

Comment:
Maybe filtering with a `.filter()` could remove 1 indentation

Reply:
BTW if you include this filter just in lines about when declaring `let drained` you'll also be able to declare results with a capacity

Reply:
Great suggestions thanks!

done: 6c5be65af717d80874e5108107b8b589d77d377c